### PR TITLE
228 enums

### LINKED ISSUE #228: (PC) consider turning the `statuses` table and `Status` model into a Rails enum on `PreservationCopy`
in our first pass at the schema, the `status` field on `PreservationCopy` was a text field.  @julianmorley was concerned that searching on this column wouldn't be performant, since text searches are more expensive than integer searches, even on indexed columns.  as such, we made the `status` col on pres copy a foreign key to a `statuses` table that holds the status text.  currently, status text is defined in the settings file, from which entries in the `statuses` table are seeded (using a method in the `Status` model).

this is getting unwieldy:  the code is aware of various statuses, and refers to them by name.  it seems unlikely that we'll add more columns to the `statuses` table, and it seems unlikely that users will add entries to the `status` settings without the code being made aware of those entries (unlike, say, endpoints, or endpoint types, or other things that might feel sort of like enums, but where there are multiple columns associated with each entry, and where the application code isn't necessarily aware of each specific entry by name in a way that's hardcoded like we're doing with status stuff).

as such, i'd propose that `status` should actually be a rails `enum`:
http://api.rubyonrails.org/v5.1/classes/ActiveRecord/Enum.html

underneath (in the DB), this will actually be an integer column (as it is right now since it's a FK to an ID in another table), and so it should still obviate julian's performance worry.  active record will do the translation that lets us refer to status by a human readable ruby symbol, as well as providing some other magical rails conveniences.

broadly, the reasons i could see to not make this change would be:
* there are actually more columns that are going to go in the `statuses` table/on the `Status` model.
  * can anyone actually think of any?
* we'll have user defined statuses that the code can work with, but that the code isn't aware of by name.
  * that sounds much more extensible and state machine-y than what we're planning on doing.
* there's some performance problem i'm not considering
  * though i'd actually think a plain int col would be more performant than a FK int col that needs to do a join.

so yeah... i vote that we do what this ticket suggests.  happy to be told i'm missing something.  i don't think the migration will be terrible:  we're still at a point where we can just build up the catalog from scratch again (and i think we're expecting to do that another time or two).  so i don't think we need to write a data migration, just a schema migration.  i think that migration is roughly:
* `preservation_copies.status_id` col gets dropped, replaced by `preservation_copies.status`, which is a plain `integer`, and not a `bigint`.
* `statuses` table gets dropped.
* `Status` model gets removed.
* `PreservationCopy` model gets an enum field called `status`, which has explicitly defined integer mappings to its status symbols (so more statuses can be safely added later).
  * `PreservationCopy` gets some new tests for its new enum (maybe some ported from the old `status_spec`, maybe not, not sure).
* other stuff that references the `Status` model gets cleaned up to use the enum (e.g `Status` seeding call gets dropped from rake db seed code, stuff that sets status on pres copy just uses the enum instead of querying for the relevant `Status` object).

### PR DESCRIPTION
**Pull request by tingulfsen**
Fixes #228 

Paired with @SaravShah 

We've removed the default status and are using 'ok' instead - a default status is awkward to implement with an enum approach.

RuboCop was complaining about a multiline expect block, wanting us to use do-end instead of curly braces. We think this looks ugly, so we turned that off for all the tests.

Note that we're now referring to statuses by strings such as 'ok'. We could potentially change this to the integers instead, or something like PreservedCopy.statuses[:ok].

MOVED TO 228-status-enums

### CODE REVIEW

File: app/models/preserved_copy.rb [hunk 1]
```diff
@@ -3,13 +3,19 @@
...
-
+  enum status: {
+    ok: 0,
+    invalid_moab: 1,
+    invalid_checksum: 2,
+    not_found_on_disk: 3,
+    expected_version_not_found_on_disk: 4,
+    fixity_check_failed: 5
+  }
+  validates :status, inclusion: { in: statuses.keys }
```
Comment (ndushay) [inline app/models/preserved_copy.rb]: I am a bit uncomfy using plain strings in the code.  Could we turn those strings into, say:  STATUS_OK here?

will something like this fly?
```
STATUS_OK = status.ok
```

Then we could have 
```
STATUS_DEFAULT = STATUS_OK
```

File: app/services/preserved_object_handler.rb [hunk 2]
```diff
@@ -65,9 +65,9 @@ def create_after_validation
     elsif PreservedObject.exists?(druid: druid)
       results << result_hash(OBJECT_ALREADY_EXISTS, 'PreservedObject')
     elsif moab_validation_errors.empty?
-      results.concat(create_db_objects(Status.default_status, true))
+      results.concat(create_db_objects("ok", true))
```
Comment (ndushay) [inline app/services/preserved_object_handler.rb]: yeah, these strings here make me nervous.  Can we get to something similar to `PreservedCopy.STATUS_OK` ?

The reasoning behind the "default" status is that we're trying to be more explicit that we're not really claiming it's okay -- we're just instantiating the damn thing and don't know its status.  This particular line was in error in its original form, as we were explicitly wanting "OK".

File: app/services/preserved_object_handler.rb [hunk 3]
```diff
@@ -81,7 +81,7 @@ def create
     elsif PreservedObject.exists?(druid: druid)
       results << result_hash(OBJECT_ALREADY_EXISTS, 'PreservedObject')
     else
-      results.concat(create_db_objects(Status.default_status))
+      results.concat(create_db_objects("ok"))
```
Comment (ndushay) [inline app/services/preserved_object_handler.rb]: this is a correct place for `PreservedCopy.STATUS_DEFAULT`

File: app/models/preserved_copy.rb [hunk 4]
```diff
@@ -3,13 +3,19 @@
...
   validates :version, presence: true
-  validates :status, presence: true
   # NOTE: size here is approximate and not used for fixity checking
   validates :size, numericality: { only_integer: true, greater_than: 0 }, allow_nil: true
-
+  enum status: {
+    ok: 0,
+    invalid_moab: 1,
+    invalid_checksum: 2,
+    not_found_on_disk: 3,
```
Comment (ndushay) [inline app/models/preserved_copy.rb]: while we're in here ... do we want to say "disk"?  Or maybe `online_moab_not_found` and `expected_version_not_found_in_moab` or some such?

File: spec/load_fixtures_helper.rb [hunk 5]
```diff
@@ -40,7 +40,7 @@ def load_fixture_moabs
       PreservedCopy.create(preserved_object_id: po.id,
                            endpoint_id: @storage_dir_to_endpoint_id[storage_dir],
                            current_version: version,
-                           status: Status.default_status)
+                           status: "ok")
```
Comment (ndushay) [inline spec/load_fixtures_helper.rb]: DEFAULT_STATUS would very much be more correct here ...

File: spec/services/preserved_object_handler_spec.rb [hunk 6]
```diff
@@ -210,10 +210,12 @@
...
         it "logs at error level" do
           expect(Rails.logger).to receive(:log).with(Logger::ERROR, version_less_than_po_msg)
-          expect(Rails.logger).to receive(:log).with(Logger::ERROR, version_less_than_pc_msg)
           expect(Rails.logger).to receive(:log).with(Logger::INFO, updated_po_db_timestamp_msg)
-          expect(Rails.logger).to receive(:log).with(Logger::INFO, updated_pc_db_obj_msg)
           expect(Rails.logger).to receive(:log).with(Logger::INFO, updated_pc_db_status_msg)
+          expect(Rails.logger).to receive(:log).with(Logger::ERROR, version_less_than_pc_msg)
+          expect(Rails.logger).to receive(:log).with(Logger::INFO, updated_pc_db_obj_msg)
+          # require 'byebug'
+          # byebug
```
Comment (ndushay) [inline spec/services/preserved_object_handler_spec.rb]: oops!  you meant to remove these comments, I bet.

**Comment by ndushay** [comment]
So should this PR be closed???

**Comment by SaravShah** [comment]
refer to #368

**Pull request closed without merging**